### PR TITLE
Product list activity

### LINKED ISSUE #18: Create activity for items page

### PR DESCRIPTION
**Pull request by vitlinda**
This pr introduces the product list activity
Closes #18

### CODE REVIEW

File: app/src/main/java/it/unibo/sc/ProductWarehouseViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    private val productName: TextView = view.findViewById(R.id.productName)
+    private val productPrice: TextView = view.findViewById(R.id.productPrice)
+
+    /**
+     * Method that bind the UI elements with [ProductWarehouse] data.
+     */
+    fun bind(productWarehouse: ProductWarehouse?) {
+        productName.text = productWarehouse?.product?.name
+        productPrice.text =
+            context.getString(R.string.euro, productWarehouse?.product?.price.toString())
```
Comment (nicolasfara) [inline app/src/main/java/it/unibo/sc/ProductWarehouseViewHolder.kt:23]: Here you are using safe call, and it's right. But what if `productWarehouse` or `product` is null? Could be better give a fallback value or manage a possible error.
For example, if `product` is `null` you could set the price string: `--` or something like that.
```
productWarehouse?.product?.price.toString() ?: "--"
```

File: app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * Class that loads pages of ProductWarehouse by page token.
+ */
+class ProductsWarehousePagingSource(private val graphQLData: ListProductsWarehouseQuery) :
+    PagingSource<String, ProductWarehouse>() {
+    override suspend fun load(params: LoadParams<String>): LoadResult<String, ProductWarehouse> {
+        return try {
+            val page = params.key
+            val response = graphQLData.getProductsWarehouse(page)
+            val items = response?.items?.toList() ?: emptyList()
+            val nextToken = response?.requestForNextResult?.variables?.get("nextToken")
```
Comment (nicolasfara) [inline app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt:19]: Also here: safe call but no fallback value

File: app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            Log.d(
+                "ProductsPagingSource",
+                "page: $page, nextToken: $nextToken, items: ${items.map { it.product.name }}"
+            )
+            LoadResult.Page(
+                data = items,
+                prevKey = null, // Only paging forward.
+                nextKey = nextToken.toString()
+            )
+        } catch (e: Exception) {
```
[the remark this answers is not in the document]
Comment (nicolasfara) [reply app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt:30]: Is there a more specific exception for handling this try/catch?

File: app/src/main/java/it/unibo/sc/viewmodel/ProductsWarehouseViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import com.amplifyframework.datastore.generated.model.ProductWarehouse
+import it.unibo.sc.ProductsWarehousePagingSource
+import it.unibo.sc.queries.ListProductsWarehouseQuery
+import kotlinx.coroutines.flow.Flow
+
+/**
+ * Class that store and manage the ProductWarehouse elements.
+ */
+class ProductsWarehouseViewModel : ViewModel() {
+    private val pageSize = 2
```
Comment (nicolasfara) [inline app/src/main/java/it/unibo/sc/viewmodel/ProductsWarehouseViewModel.kt:17]: Constants should be placed in a companion object

**Review by nicolasfara — CHANGES_REQUESTED** [review]

File: app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (vitlinda) [reply app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt:30]: nope

File: app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt (hunk 2 above)
Comment (vitlinda) [reply app/src/main/java/it/unibo/sc/ProductsWarehousePagingSource.kt:19]: the fallback value in this case would be "null" which is an accepted value (as in grapqhl a nextToken = null means last page) so elvis operator would be redundant

**Review by nicolasfara — APPROVED** [review]

**Pull request merged**